design/fips-compliance.md

```diff
@@ -0,0 +1,199 @@
+# FIPS compliance for conjur-authn-k8s-client
+
+## Background
+
+The National Institute of Standards and Technology (NIST) issued the [FIPS
+ 140-2](https://csrc.nist.gov/publications/detail/fips/140/2/final) 
+Publication Series to coordinate the requirements and standards for cryptography 
+modules that include both hardware and software components.
+
+By making the conjur-authn-k8s-client FIPS compliant, our costumer will gain a 
+cryptographic standard to protect their unclassified, but sensitive data.
+
+## Useful Links
+
+| **Name**                        | **Link**                                            |
+|---------------------------------|-----------------------------------------------------|
+| FIPS compliant crypto in golang | https://kupczynski.info/2019/12/15/fips-golang.html |
+
+## Solution
+
+We looked into 3 options for making out product FIPS compliant:
+
+1. Use cgo to call out to an existing certified library
+   
+    - This alternative disqualified due to its complexity and consuming time
+     and effort as we will have to replace any usage of crypto.
+     
+1. Use [RHEL go toolchain](https://developers.redhat.com/blog/2019/06/24/go-and-fips-140-2-on-red-hat-enterprise-linux/).
+   - RHEL takes the ownership on the encryption, and bridges between the
+    Golang encryption and OpenSSL.
+   - What needs to be done is build the Golang project on subscribed RHEL
+    machine/UBI container with the relevant go-toolset installed on it. then, 
+    we can take the compiled binary and copy it into the containers as usual.
+   - A big advantage of this approach is that we are aligning with Red Hat
+    which is one of the big players in the field of certifications, enterprise, 
+    and federal market.
+   - The pipeline should use and subscribe RHEL / UBI on-the-fly.
+   
+1. Use BoringSSL based crypto
+   Use [Google’s fork of Golang](https://github.com/golang/go/blob/dev.boringcrypto.go1.12/misc/boring/README.md) 
+   that uses dev.boringcrypto as its crypto engine. dev.boringcrypto wraps BoringSSL which is FIPS compliant.
+   
+Looking into the options above, it was clear that the best option is BoringSSL.
+Using cgo (option 1) was disqualified due to its complexity and consuming time
+ and effort as we will have to replace any usage of crypto.
+ 
+Using the RHEL go toolchain has its infrastructure complexity as we will need
+ to build a new pipeline for building the project on a subscribed RHEL machine.
+ 
+Luckily, Google maintains the BoringSSL which is very easy to use and gives
+ us FIPS compliancy with minimal work. All that needs to be done is:
+ - Replace the base image from `golang` to the corresponding `goboring/golang
+ `. For example, `golang:1.12` should turn into `goboring/golang:1.12.17b4`
+ - Enable cgo by setting `CGO_ENABLED=1` in the env
+ - Verify that the `authenticator` binary uses the BoringCrypto libraries by
```

Comment:
And if it doesn't? Or if a project doesn't, what is the amount of work that needs to be done?

Reply:
first, it does :)

Second, if it doesn't then we missed some step in making the project FIPS compliant. This step verifies that everything went well and the project is FIPS compliant.

Reply:
it's not a matter of work to do on the project, only follow the steps of making a project FIPS-Compliant